src/tty/unix.rs

```diff
@@ -15,59 +15,79 @@
 //! tty related functionality
 //!
 
-use crate::tty::EventedReadWrite;
+use crate::tty::{EventedReadWrite, ChildInfo};
 use crate::term::SizeInfo;
 use crate::display::OnResize;
 use crate::config::{Config, Shell};
 use crate::cli::Options;
 use mio;
 
-use libc::{self, c_int, pid_t, winsize, SIGCHLD, TIOCSCTTY, WNOHANG};
+use libc::{self, c_int, winsize, SIGCHLD, TIOCSCTTY};
 
-use std::os::unix::io::{FromRawFd, RawFd};
+use std::os::unix::io::{FromRawFd};
 use std::fs::File;
 use std::os::unix::process::CommandExt;
-use std::process::{Command, Stdio};
+use std::process::{Command, Stdio, Child};
 use std::ffi::CStr;
 use std::ptr;
 use mio::unix::EventedFd;
-use std::io;
+use std::io::{self, Write, Read};
 use std::os::unix::io::AsRawFd;
+use std::mem;
 
+/// Get the current value of errno
+fn errno() -> c_int {
+    ::errno::errno().0
+}
 
-/// Process ID of child process
-///
-/// Necessary to put this in static storage for `sigchld` to have access
-pub static mut PID: pid_t = 0;
+fn set_errno(value: c_int) {
+    ::errno::set_errno(::errno::Errno(value));
+}
 
-/// Exit flag
+/// Self-pipe
 ///
-/// Calling exit() in the SIGCHLD handler sometimes causes opengl to deadlock,
-/// and the process hangs. Instead, this flag is set, and its status can be
-/// checked via `process_should_exit`.
-static mut SHOULD_EXIT: bool = false;
-
-extern "C" fn sigchld(_a: c_int) {
-    let mut status: c_int = 0;
-    unsafe {
-        let p = libc::waitpid(PID, &mut status, WNOHANG);
-        if p < 0 {
-            die!("Waiting for pid {} failed: {}\n", PID, errno());
+/// We write any signals into the pipe, and the event loop wakes up and reads them.
+static mut SIG_PIPE_WRITE: Option<File> = None;
```

Comment:
This is an `Optional` because it gets initialized when spawning the child - adding a little sync to this would probably work out nicely when we eventually get around to multiple windows (like #1589).

Reply:
I feel like we should either handle a `None` value here or choose a different way to store this, otherwise I'm not sure which benefit we get from having an Option if we always crash anyways when it's `None`.

Reply:
It might be my weak rust-fu - I used `Option` just because it has to be initialized dynamically.

Reply:
Turns out all the alternatives are worse, currently. The only other legit way to do this on `stable` is `lazy_static!`, and that's even more overkill that `Option`.